### ISSUE TITLE
"Module 'com.google.guava:guava' has been rejected" error during unit tests compilation

### ISSUE THREAD

**Issue by MyDogTom** [jjohannes/missing-metadata-guava#4]
Hi. 
I have an Android project where some modules have explicit dependency on Guava library just to workaround resolution conflicts. Guava itself is coming as transitive dependency from WorkManager, ExoPlayer, and probably some other libraries.

I've applied this plugin and was able to remove explicit dependency on Guava library. Application compiles. However unit tests compilation (`assembleDebugUnitTest`) fails with the error

```
Could not determine the dependencies of task ':settings:processDebugUnitTestResources'.
> Could not resolve all task dependencies for configuration ':settings:debugUnitTestRuntimeClasspath'.
   > Could not resolve com.google.guava:guava:27.0.1-jre.
     Required by:
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:resources:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:sandbox:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:utils:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:plugins-maven-dependency-resolver:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:shadows-framework:4.7.2 > org.robolectric:nativeruntime:4.7.2
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:listenablefuture:1.0' also provided by [com.google.guava:listenablefuture:1.0(runtime), com.google.guava:guava:27.0.1-jre(androidRuntime)]
   > Could not resolve com.google.guava:listenablefuture:1.0.
     Required by:
         project :settings > project :base_ui > androidx.work:work-runtime:2.5.0
         project :settings > project :navigation > androidx.browser:browser:1.3.0
         project :settings > project :navigation > androidx.browser:browser:1.3.0 > androidx.concurrent:concurrent-futures:1.0.0
      > Module 'com.google.guava:listenablefuture' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:listenablefuture:1.0' also provided by [com.google.guava:guava:27.0.1-jre(androidRuntime), com.google.guava:guava:27.1-android(runtime)]
   > Could not resolve com.google.guava:guava:27.1-android.
     Required by:
         project :settings > project :base-api > com.google.android.exoplayer:exoplayer:2.13.0 > com.google.android.exoplayer:exoplayer-core:2.13.0 > com.google.android.exoplayer:exoplayer-common:2.13.0
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:listenablefuture:1.0' also provided by [com.google.guava:listenablefuture:1.0(runtime), com.google.guava:guava:27.0.1-jre(androidRuntime)]
```

What confuses me is that If I don't apply `de.jjohannes.missing-metadata-guava`, then unit test compilation for this module passes. (In that case application compilation fails as I removed explicit Guava dependency). Compilation also passes if I add explicit dependency `testImplementation "com.google.guava:guava:27.1-android"` to this module.

Do I miss something? Do I have to use explicit `testImplementation` Guava dependency and it's the recommended way to go?

I attach result of `./gradlew myModule:dependencyInsight --configuration debugUnitTestRuntimeClasspath --dependency :guava:`. Hope it gives you extra information.

When plugin is applied and `assembleDebugUnitTest` fails.

```
com.google.guava:guava:27.0.1-jre FAILED
   Failures:
      - Could not resolve com.google.guava:guava:27.0.1-jre.
          - Module 'com.google.guava:guava' has been rejected:
               Cannot select module with conflict on capability 'com.google.guava:listenablefuture:1.0' also provided by [com.google.guava:listenablefuture:1.0(runtime), com.google.guava:guava:27.0.1-jre(androidRuntime)]

com.google.guava:guava:27.0.1-jre FAILED
+--- org.robolectric:nativeruntime:4.7.2
|    \--- org.robolectric:shadows-framework:4.7.2
|         \--- org.robolectric:robolectric:4.7.2
|              \--- project :settings
|                   \--- project :settings (*)
+--- org.robolectric:plugins-maven-dependency-resolver:4.7.2
|    \--- org.robolectric:robolectric:4.7.2 (*)
+--- org.robolectric:resources:4.7.2
|    +--- org.robolectric:robolectric:4.7.2 (*)
|    \--- org.robolectric:shadows-framework:4.7.2 (*)
+--- org.robolectric:sandbox:4.7.2
|    +--- org.robolectric:robolectric:4.7.2 (*)
|    \--- org.robolectric:junit:4.7.2
|         \--- org.robolectric:robolectric:4.7.2 (*)
\--- org.robolectric:utils:4.7.2
     +--- org.robolectric:robolectric:4.7.2 (*)
     +--- org.robolectric:shadows-framework:4.7.2 (*)
     +--- org.robolectric:resources:4.7.2 (*)
     +--- org.robolectric:sandbox:4.7.2 (*)
     +--- org.robolectric:utils-reflector:4.7.2
     |    +--- org.robolectric:robolectric:4.7.2 (*)
     |    +--- org.robolectric:shadows-framework:4.7.2 (*)
     |    +--- org.robolectric:junit:4.7.2 (*)
     |    \--- org.robolectric:sandbox:4.7.2 (*)
     +--- org.robolectric:plugins-maven-dependency-resolver:4.7.2 (*)
     \--- org.robolectric:shadowapi:4.7.2
          +--- org.robolectric:shadows-framework:4.7.2 (*)
          +--- org.robolectric:junit:4.7.2 (*)
          \--- org.robolectric:sandbox:4.7.2 (*)

com.google.guava:guava:27.1-android FAILED
   Failures:
      - Could not resolve com.google.guava:guava:27.1-android. (already reported)

com.google.guava:guava:27.1-android FAILED
\--- com.google.android.exoplayer:exoplayer-common:2.13.0
     +--- com.google.android.exoplayer:exoplayer-core:2.13.0
     |    +--- com.google.android.exoplayer:exoplayer:2.13.0
     |    |    \--- project ... // other modules from the project
     |    +--- com.google.android.exoplayer:exoplayer-ui:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-dash:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-hls:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-smoothstreaming:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    \--- com.google.android.exoplayer:exoplayer-transformer:2.13.0
     |         \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     \--- com.google.android.exoplayer:exoplayer-extractor:2.13.0
          \--- com.google.android.exoplayer:exoplayer-core:2.13.0 (*)

(*) - dependencies omitted (listed previously)
```

When plugin is not applied and `assembleDebugUnitTest` passes

```
com.google.guava:guava:27.1-android
   variant "runtime" [
      org.gradle.status                              = release (not requested)
      org.gradle.usage                               = java-runtime
      org.gradle.libraryelements                     = jar (not requested)
      org.gradle.category                            = library (not requested)

      Requested attributes not found in the selected variant:
         com.android.build.api.attributes.BuildTypeAttr = debug
         org.gradle.jvm.environment                     = android
         org.jetbrains.kotlin.platform.type             = androidJvm
   ]
   Selection reasons:
      - By conflict resolution : between versions 27.1-android and 27.0.1-jre

com.google.guava:guava:27.1-android
\--- com.google.android.exoplayer:exoplayer-common:2.13.0
     +--- com.google.android.exoplayer:exoplayer-core:2.13.0
     |    +--- com.google.android.exoplayer:exoplayer:2.13.0
     |    |    \--- project ... // other modules from the project
     |    +--- com.google.android.exoplayer:exoplayer-ui:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-dash:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-hls:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-smoothstreaming:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    \--- com.google.android.exoplayer:exoplayer-transformer:2.13.0
     |         \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     \--- com.google.android.exoplayer:exoplayer-extractor:2.13.0
          \--- com.google.android.exoplayer:exoplayer-core:2.13.0 (*)

com.google.guava:guava:27.0.1-jre -> 27.1-android
+--- org.robolectric:nativeruntime:4.7.2
|    \--- org.robolectric:shadows-framework:4.7.2
|         \--- org.robolectric:robolectric:4.7.2
|              \--- project :settings
|                   \--- project :settings (*)
+--- org.robolectric:plugins-maven-dependency-resolver:4.7.2
|    \--- org.robolectric:robolectric:4.7.2 (*)
+--- org.robolectric:resources:4.7.2
|    +--- org.robolectric:robolectric:4.7.2 (*)
|    \--- org.robolectric:shadows-framework:4.7.2 (*)
+--- org.robolectric:sandbox:4.7.2
|    +--- org.robolectric:robolectric:4.7.2 (*)
|    \--- org.robolectric:junit:4.7.2
|         \--- org.robolectric:robolectric:4.7.2 (*)
\--- org.robolectric:utils:4.7.2
     +--- org.robolectric:robolectric:4.7.2 (*)
     +--- org.robolectric:shadows-framework:4.7.2 (*)
     +--- org.robolectric:resources:4.7.2 (*)
     +--- org.robolectric:sandbox:4.7.2 (*)
     +--- org.robolectric:utils-reflector:4.7.2
     |    +--- org.robolectric:robolectric:4.7.2 (*)
     |    +--- org.robolectric:shadows-framework:4.7.2 (*)
     |    +--- org.robolectric:junit:4.7.2 (*)
     |    \--- org.robolectric:sandbox:4.7.2 (*)
     +--- org.robolectric:plugins-maven-dependency-resolver:4.7.2 (*)
     \--- org.robolectric:shadowapi:4.7.2
          +--- org.robolectric:shadows-framework:4.7.2 (*)
          +--- org.robolectric:junit:4.7.2 (*)
          \--- org.robolectric:sandbox:4.7.2 (*)

(*) - dependencies omitted (listed previously)
```

When plugin is applied and explicit Guava dependencies is added via `testImplementation`. `assembleDebugUnitTest` passes.

```
com.google.guava:guava:27.1-android
   variant "runtime" [
      org.gradle.jvm.version                         = 6 (not requested)
      org.gradle.jvm.environment                     = android
      org.gradle.status                              = release (not requested)
      org.gradle.usage                               = java-runtime
      org.gradle.libraryelements                     = jar (not requested)
      org.gradle.category                            = library (not requested)

      Requested attributes not found in the selected variant:
         com.android.build.api.attributes.BuildTypeAttr = debug
         org.jetbrains.kotlin.platform.type             = androidJvm
   ]
   Selection reasons:
      - By conflict resolution : between versions 27.1-android and 27.0.1-jre

com.google.guava:guava:27.1-android
+--- project :settings
|    \--- project :settings (*)
\--- com.google.android.exoplayer:exoplayer-common:2.13.0
     +--- com.google.android.exoplayer:exoplayer-core:2.13.0
     |    +--- com.google.android.exoplayer:exoplayer:2.13.0
     |    |    \--- project ... // other modules from the project
     |    +--- com.google.android.exoplayer:exoplayer-dash:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-hls:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-smoothstreaming:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    +--- com.google.android.exoplayer:exoplayer-transformer:2.13.0
     |    |    \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     |    \--- com.google.android.exoplayer:exoplayer-ui:2.13.0
     |         \--- com.google.android.exoplayer:exoplayer:2.13.0 (*)
     \--- com.google.android.exoplayer:exoplayer-extractor:2.13.0
          \--- com.google.android.exoplayer:exoplayer-core:2.13.0 (*)

com.google.guava:guava:27.0.1-jre -> 27.1-android
+--- org.robolectric:nativeruntime:4.7.2
|    \--- org.robolectric:shadows-framework:4.7.2
|         \--- org.robolectric:robolectric:4.7.2
|              \--- project :settings
|                   \--- project :settings (*)
+--- org.robolectric:plugins-maven-dependency-resolver:4.7.2
|    \--- org.robolectric:robolectric:4.7.2 (*)
+--- org.robolectric:resources:4.7.2
|    +--- org.robolectric:robolectric:4.7.2 (*)
|    \--- org.robolectric:shadows-framework:4.7.2 (*)
+--- org.robolectric:sandbox:4.7.2
|    +--- org.robolectric:robolectric:4.7.2 (*)
|    \--- org.robolectric:junit:4.7.2
|         \--- org.robolectric:robolectric:4.7.2 (*)
\--- org.robolectric:utils:4.7.2
     +--- org.robolectric:robolectric:4.7.2 (*)
     +--- org.robolectric:shadows-framework:4.7.2 (*)
     +--- org.robolectric:resources:4.7.2 (*)
     +--- org.robolectric:sandbox:4.7.2 (*)
     +--- org.robolectric:utils-reflector:4.7.2
     |    +--- org.robolectric:robolectric:4.7.2 (*)
     |    +--- org.robolectric:shadows-framework:4.7.2 (*)
     |    +--- org.robolectric:junit:4.7.2 (*)
     |    \--- org.robolectric:sandbox:4.7.2 (*)
     +--- org.robolectric:plugins-maven-dependency-resolver:4.7.2 (*)
     \--- org.robolectric:shadowapi:4.7.2
          +--- org.robolectric:shadows-framework:4.7.2 (*)
          +--- org.robolectric:junit:4.7.2 (*)
          \--- org.robolectric:sandbox:4.7.2 (*)

(*) - dependencies omitted (listed previously)
```

**Comment by jjohannes** [comment]
Hi @MyDogTom. The plugin is adding the `com.google.guava:listenablefuture` [capability](https://docs.gradle.org/current/userguide/component_capabilities.html) to Guava, because Guava contains that feature directly nowadays. And it removes a "hack" Guava is doing to upgrade to an empty `listenablefuture.jar` with version `9999.0-empty-to-avoid-conflict-with-guava` so that you do not get that empty file on your classpath.

If you have something else depending on `listenablefuture` directly - e.g. `androidx.concurrent:concurrent-futures` - you will get a conflict (which you are seeing). 

You need to add a [resolution strategy for the capability](https://docs.gradle.org/current/userguide/dependency_capability_conflict.html#sub:selecting-between-candidates) to resolve the conflict:
```
  configurations.all {
        resolutionStrategy.capabilitiesResolution.withCapability("com.google.guava:listenablefuture") {
            selectHighestVersion()
        }
    }
```

**Comment by MyDogTom** [comment]
Thanks for your help! 
I tried your snippet, but still received an error.
```
Could not determine the dependencies of task ':settings:mergeDebugUnitTestResources'.
> Could not resolve all task dependencies for configuration ':settings:debugUnitTestRuntimeClasspath'.
   > Could not resolve com.google.guava:guava:27.0.1-jre.
     Required by:
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:resources:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:sandbox:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:utils:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:plugins-maven-dependency-resolver:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:shadows-framework:4.7.2 > org.robolectric:nativeruntime:4.7.2
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:listenablefuture:1.0' also provided by [com.google.guava:listenablefuture:1.0(runtime), com.google.guava:guava:27.0.1-jre(androidRuntime)]
   > Could not resolve com.google.guava:listenablefuture:1.0.
     Required by:
         project :settings > project :base_ui > androidx.work:work-runtime:2.5.0
         project :settings > project :navigation > androidx.browser:browser:1.3.0
         project :settings > project :navigation > androidx.browser:browser:1.3.0 > androidx.concurrent:concurrent-futures:1.0.0
      > Module 'com.google.guava:listenablefuture' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:listenablefuture:1.0' also provided by [com.google.guava:guava:27.0.1-jre(androidRuntime), com.google.guava:guava:27.1-android(runtime)]
   > Could not resolve com.google.guava:guava:27.1-android.
     Required by:
         project :settings > project :base-api > com.google.android.exoplayer:exoplayer:2.13.0 > com.google.android.exoplayer:exoplayer-core:2.13.0 > com.google.android.exoplayer:exoplayer-common:2.13.0
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:listenablefuture:1.0' also provided by [com.google.guava:listenablefuture:1.0(runtime), com.google.guava:guava:27.0.1-jre(androidRuntime)]
```

Then I thought that probably `selectHighestVersion()` is not able to choose between `com.google.guava:listenablefuture:1.0`, `com.google.guava:guava:27.0.1-jre` and `com.google.guava:guava:27.1-android`. I received these three by printing `candidate.id.displayName`.

Based on that assumption I excluded `com.google.guava:listenablefuture:1.0` from resolution to ensure that it's never selected.

```
configuration.resolutionStrategy.capabilitiesResolution
    .withCapability("com.google.guava:listenablefuture") { resolution ->
        val onlyGuava = resolution.candidates.filter {
            it.id.displayName.startsWith("com.google.guava:guava")
        }
        val selected = onlyGuava.maxByOrNull {
            VersionNumber.parse(it.id.displayName.substringAfterLast(":"))
        }
        selected?.let {
            resolution.select(it)
        } ?: throw IllegalStateException("can't find resolution")
    }
```

this time the error was different

```
Could not determine the dependencies of task ':settings:mergeDebugUnitTestResources'.
> Could not resolve all task dependencies for configuration ':settings:debugUnitTestRuntimeClasspath'.
   > Could not resolve com.google.guava:guava:27.0.1-jre.
     Required by:
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:resources:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:sandbox:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:utils:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:plugins-maven-dependency-resolver:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:shadows-framework:4.7.2 > org.robolectric:nativeruntime:4.7.2
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.collections:google-collections:27.1' also provided by [com.google.guava:guava:27.0.1-jre(androidRuntime)]
   > Could not resolve com.google.guava:listenablefuture:1.0.
     Required by:
         project :settings > project :base_ui > androidx.work:work-runtime:2.5.0
         project :settings > project :navigation > androidx.browser:browser:1.3.0
         project :settings > project :navigation > androidx.browser:browser:1.3.0 > androidx.concurrent:concurrent-futures:1.0.0
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.collections:google-collections:27.1' also provided by [com.google.guava:guava:27.0.1-jre(androidRuntime)]
   > Could not resolve com.google.guava:guava:27.1-android.
     Required by:
         project :settings > project :base-api > com.google.android.exoplayer:exoplayer:2.13.0 > com.google.android.exoplayer:exoplayer-core:2.13.0 > com.google.android.exoplayer:exoplayer-common:2.13.0
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.collections:google-collections:27.1' also provided by [com.google.guava:guava:27.0.1-jre(androidRuntime)]
```

Then I added resolution  for `com.google.collections:google-collections` and `com.google.guava:guava` (next error I received) using `selectHighestVersion`.

```
configuration.resolutionStrategy
    .capabilitiesResolution
    .withCapability("com.google.collections:google-collections") { resolution ->
        resolution.selectHighestVersion()
    }
configuration.resolutionStrategy
    .capabilitiesResolution
    .withCapability("com.google.guava:guava") { resolution ->
        resolution.selectHighestVersion()
    }
```

This time compilation succeeded. 

To be honest the final result looks a bit overcomplicated, but I guess it's the way to go, right?

**Comment by jjohannes** [comment]
That you also need to do `com.google.collections:google-collections` makes sense. But you should not need to do something for `com.google.guava:guava`. Also, the more complicated resolution you did should not be necessary - `selectHighestVersion()` should be enough.

`com.google.guava:guava:27.0.1-jre` and `com.google.guava:guava:27.1-android` should usually not turn up in the capability conflict. They are the same module (only with different versions).

Maybe there is indeed an issue with the plugin. I will try to reproduce the problem.

Which Gradle and Android versions are you using?

**Comment by MyDogTom** [comment]
Gradle 7.3.3 and AGP 7.0.4.

**Comment by MyDogTom** [comment]
>But you should not need to do something for com.google.guava:guava.

btw, If I don't include resolution for `com.google.guava:guava` then I see the following error

```
> Could not resolve all task dependencies for configuration ':settings:debugUnitTestRuntimeClasspath'.
   > Could not resolve com.google.guava:guava:27.0.1-jre.
     Required by:
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:resources:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:sandbox:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:utils:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:plugins-maven-dependency-resolver:4.7.2
         project :settings > org.robolectric:robolectric:4.7.2 > org.robolectric:shadows-framework:4.7.2 > org.robolectric:nativeruntime:4.7.2
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:guava:27.1-android' also provided by [com.google.guava:guava:27.0.1-jre(androidRuntime)]
   > Could not resolve com.google.guava:listenablefuture:1.0.
     Required by:
         project :settings > project :base_ui > androidx.work:work-runtime:2.5.0
         project :settings > project :navigation > androidx.browser:browser:1.3.0
         project :settings > project :navigation > androidx.browser:browser:1.3.0 > androidx.concurrent:concurrent-futures:1.0.0
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:guava:27.1-android' also provided by [com.google.guava:guava:27.0.1-jre(androidRuntime)]
   > Could not resolve com.google.guava:guava:27.1-android.
     Required by:
         project :settings > project :base-api > com.google.android.exoplayer:exoplayer:2.13.0 > com.google.android.exoplayer:exoplayer-core:2.13.0 > com.google.android.exoplayer:exoplayer-common:2.13.0
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'com.google.guava:guava:27.1-android' also provided by [com.google.guava:guava:27.0.1-jre(androidRuntime)]
```

**Comment by jjohannes** [comment]
Sorry, I was mistaken. I forgot that the plugin actually adds the conflict resolution rules already. So it should just work and there should be no need to add resolution rules for capability conflicts.

I tried to reproduce your issue and added a test with the Android plugin, but had no success. Would you be able to share a reproducer?

**Comment by MyDogTom** [comment]
Hi @jjohannes, 
I added a simple project that produces similar error https://github.com/MyDogTom/GuavaReproduce

Failure happens after executing `./gradlew app:testDebugUnitTest`.

**Comment by MyDogTom** [comment]
I've just noticed that if I bump ExoPlayer dependency to 2.16.0, then the issue is gone. With ExoPlayer 2.15.1 version the build still fails. I've quickly scanned [ExoPlayer 2.16.0 changelog](https://github.com/google/ExoPlayer/blob/release-v2/RELEASENOTES.md#2160-2021-11-04). The only change that seems to be related is 
>Upgrade the WorkManager extension to depend on androidx.work:work-runtime:2.7.0. 

Still I don't have enough knowledge to find a complete explanation :(

**Comment by jjohannes** [comment]
Thanks for the reproduced @MyDogTom. I was finally able to track down the issue and fix it in 494bd4627c887f02993a5d0dc4e9165074375211.

I don't know though, why upgrading ExoPlayer fixed it and why it exists in the first place. It must be related to in which order Gradle does things and this might also be a "bug" in Gradle itself (I am not sure).

Thanks again for reporting this and your patience. I will release a `31.1.1` version of the plugin with the fix.

**Comment by jeff-huston** [comment]
This fixed a build issue we were having, too - thanks @MyDogTom and @jjohannes!
